opal/util/json/opal_json.h

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (c) 2024      Amazon.com, Inc. or its affiliates.
+ *                         All Rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+/** @file:
+ *
+ * Utility data structures and functions to parse JSON from a string or file.
+ * Note: The internal 3rd-party code should NOT be directly used. Implement the
+ * functionality here if necessary.
+ *
+ * Below is a trivial example to parse a JSON file with the content {"key": "val"}.
+
+int ret = OPAL_SUCCESS;
+const opal_json_t *json = NULL, *item = NULL;
+char *val;
+size_t len;
+
+ret = opal_json_load_file("my_file.json", &json); // Parse the file content into json
+if (OPAL_SUCCESS != ret) {
+     goto out;
+}
+
+ret = opal_json_get_key(json, "key", &item); // Get the item pointed to by "key"
+if (OPAL_SUCCESS != ret) {
+     goto out;
+}
+
+ret = opal_json_read_string(item, (const char **) &val, &len); // Read the string value and length
+if (OPAL_SUCCESS != ret) {
+    goto out;
+}
+
+out:
+
+// Cleanup
+opal_json_free(&json);
+opal_json_free(&item);
+
+ */
+
+#ifndef OPAL_JSON_H
+#define OPAL_JSON_H
+
+#include "opal/mca/mca.h"
+
+BEGIN_C_DECLS
+
+typedef enum {
+    OPAL_JSON_NULL,
+    OPAL_JSON_OBJECT,
+    OPAL_JSON_ARRAY,
+    OPAL_JSON_INTEGER,
+    OPAL_JSON_DOUBLE,
+    OPAL_JSON_BOOL,
+    OPAL_JSON_STRING,
+    OPAL_JSON_TYPE_COUNT,
+} opal_json_type;
+
+struct opal_json_t {
+    opal_json_type type;
+};
+typedef struct opal_json_t opal_json_t;
+
+/**
+ * Load JSON from a string.
+ *
+ * @param[in]   str     A JSON string
+ * @param[in]   len     First bytes to parse
+ * @param[out]  json    Output JSON object if the string is valid
+ *
+ * @returns     OPAL_SUCCESS if the string is parsed successfully
+ *              OPAL_ERROR otherwise
+ */
+OPAL_DECLSPEC int opal_json_load(const char *str, const size_t len, const opal_json_t **json);
+
+/**
+ * Load JSON from a file.
+ *
+ * @param[in]   filename    File name
+ * @param[out]  json        Output JSON object if the file valid
+ *
+ * @returns     OPAL_SUCCESS if the file is read and parsed successfully
+ *              OPAL_ERROR otherwise
+ */
+OPAL_DECLSPEC int opal_json_load_file(const char *filename, const opal_json_t **json);
+
+/**
+ * Free JSON resources
+ *
+ * @param[in]  json     Pointer to JSON object to free
```

Comment:
Is `json` set to NULL upon return?  The param is marked `in`, but you pass a `**`.

Reply:
Nope `json` is not modified upon return.

The `**` is due to the fact of that `json` being used(and likely declared) as `const *` everywhere - here we need a double pointer to tell the compiler that we are not modifying `json`, but rather a pointer to it.